winthrop/books/migrations/0015_plum_to_figgy.py

```diff
@@ -0,0 +1,138 @@
+                # if short id lookup fails, map by index
+                db_canvas.uri = newmanif.sequences[0].canvases[index].id
+
+            db_canvas.save()
+
+            # NOTE: not updating other canvas metadata here; use
+            # existing import/update script once urls have been migrated
+
+        # save the manifest with the new figgy url
+        db_manif.save()
+
+        # use standard import/update logic to update everything else
+        mnf_importer.import_manifest(newmanif, newmanif.id)
+
+        imgs = []
+        figgy_imgs = []
+        for ann in Annotation.objects.filter(uri__contains=plum_uri):
+            # annotations have a db-association with canvas objects,
+            # but also reference canvas URI as annotation target and
+            # and IIIF image url in image selection
+            old_canvas_id = ann.uri.rsplit('/')[-1]
+
+            # update annotated canvas uri
+            new_canvas = canvas_by_id[old_canvas_id]
+            ann.uri = new_canvas.id
+
+            # use piffle iiif image handling to update iiif server and
+            # image id without modifying other image parameters
+            # - parse image selection uri as iiif image
+            if 'image_selection' in ann.extra_data:
+                img = iiif.IIIFImageClient.init_from_url(
+                    ann.extra_data['image_selection']['uri']
+                )
+                # - load iiif image service as iiifimage
+                figgy_img = iiif.IIIFImageClient(
+                    *new_canvas.images[0].resource.service.id.rsplit('/', 1)
+                    )
+                # update existing iiif image url with new server & image id
+                img.api_endpoint = figgy_img.api_endpoint
+                img.image_id = figgy_img.image_id
+                # image uri is stored in two places, so update both
+                ann.extra_data['image_selection']['uri'] = str(img)
+                ann.extra_data['image_selection']['src'] = str(img)
+
+            ann.save()
+
+    # as a sanity check, in case anything went wrong -
+    # check for any unmigrated plum uris and warn if any are found
+    # NOTE: This lines are longer than PEP8, but splitting them results in
+    # more confusion than not.
+    unmigrated = {
+        'manifest': Manifest.objects.filter(uri__contains='plum.princeton').count(),
+        'canvas': Canvas.objects.filter(uri__contains='plum.princeton').count(),
+        'intervention': Annotation.objects.filter(models.Q(uri__contains='plum.princeton') |
+                                                  models.Q(extra_data__contains='plum.princeton'))
+                                                  .count()
+    }
+    if any(unmigrated.values()):
+        print('Found unmigrated content: %(manifest)d manifests, '
+              '%(canvas)d canvases, %(intervention)d interventions' %
+              unmigrated)
+        # This will trigger an error
+        assert False
+
```

Comment:
couldn't we just `raise` a generic `Exception` with this text?

Reply:
Agreed, seems better to raise an exception. Maybe could be a `ValidationError` ? (not strictly validation, but was the closest I could find on a quick look)

Reply:
I think `ValidationError` is what occurred to me when @thatbudakguy suggested an Exception.